content/blog/2025-12-07-avoid-consecutive-repartitions.md

```diff
@@ -0,0 +1,428 @@
+
+When I found out I would be on a team focusing on query engines and execution, I was excited but horrified. "Query engines?" From my experience, I typed SQL queries into pgAdmin and got responses without knowing the dark magic that happened under the hood.
+
+With what seemed like an impossible task at hand, I began my favorite few months of learning.
+
+---
+
+## **Starting Out**
+
+I am no expert in databases or any of their subsystems, but I am someone who recently began learning about them. These are some tips I find useful when first starting.
+
+### Build a Foundation
+
+The first thing I did, which I highly recommend, was watch Andy Pavlo's [Intro To Database Systems course](https://15445.courses.cs.cmu.edu/fall2025/). This laid a great foundation for understanding how a database works from end to end at a high level. It touches on topics ranging from file formats to query optimization, and it was helpful to have a general context for the whole system before diving deep into a single sector.
+
+### Narrow Your Scope
+
+The next crucial step is to pick your niche and stick to it. Database systems are so vast that trying to tackle the whole beast at once is a lost cause. If you want to effectively contribute to this space, you need to deeply understand the system you are working on, and you will have much better luck narrowing your scope.
+
+When learning about the entire database stack at a high level, note what parts stick out as particularly interesting. For me, this focus is on query engines, more specifically, the physical planner and optimizer.
+
+### A "Slow" Start
+
+The final piece of advice when starting, and I sound like a broken record, is to take your time to learn. This is not an easy sector of software to jump into; it will pay dividends to slow down, fully understand the system, and why it is designed the way it is.
+
+When making your first contributions to an open-source project, start very small but go as deep as you can. Don't leave any stone unturned. I did this by looking for simpler issues, such as formatting or simple bug fixes, and stepping through the entire data flow that relates to the issue, noting what each component is responsible for.
+
+This will give you familiarity with the codebase and using your tools, like your debugger, within the project.
+
+<div class="text-center">
+<img
+  src="/blog/images/avoid-consecutive-repartitions/noot_noot_database_meme.png"
+  width="50%"
+  class="img-responsive"
+  alt="Noot Noot Database Meme"
+/>
+</div>
+<br>
+
+Now that we have some general knowledge of database internals, a niche or subsystem we want to dive deeper into, and the mindset for acquiring knowledge before contributing, let's start with our first core issue.
+
+---
+
+## **Intro to Datafusion**
```

Comment:
A nit is that the formal name of the project is `DataFusion`  (capital `F`) so it would be nice t use that form in the text